packages/web-components/src/components/va-modal/va-modal.tsx

```diff
@@ -0,0 +1,310 @@
+
+  /**
+   * If true, analytics event won't be fired
+   */
+  @Prop() disableAnalytics?: boolean = false;
+
+  /**
+   * Title/header text for the modal
+   */
+  @Prop() modalTitle?: string;
+
+  /**
+   * Hide the close button that's normally in the top right
+   */
+  @Prop() hideCloseButton?: boolean = false;
+
+  /**
+   * Selector to explicitly specify which element should receive
+   * focus when the modal is open, if the initially focused element
+   * is not the first focusable element in the document
+   */
+  @Prop() initialFocusSelector?: string;
+
+  /**
+   * Primary button text and action
+   */
+  @Prop() primaryButton?: any;
+
+  /**
+   * Secondary button text and action
+   */
+  @Prop() secondaryButton?: any;
+
+  /*
+   * Style of modal alert - info, error, success, warning
+   */
+  @Prop() status?: 'continue' | 'error' | 'info' | 'success' | 'warning';
+
+  /**
+   * If the modal is visible or not
+   */
+  @Prop() visible: boolean = false;
+
+  @Listen('click')
+  handleClick(e) {
+    if (!this.clickToClose) return;
+
+    // event.target is always the shadow host
+    // event.composedPath()[0] returns the node clicked when shadow root is open
+    if (this.visible && e.composedPath()[0] === this.el) {
+      this.handleClose(e);
+    }
+  }
+
+  @Listen('keydown', { target: 'window' })
+  handleKeyDown(e: KeyboardEvent) {
+    if (!this.visible) return;
+
+    const keyCode = e.key;
+    if (keyCode === 'Escape') {
+      this.handleClose(e);
+    }
+  }
+
+  @Watch('visible')
+  watchVisibleHandler() {
+    // This is a workaround for determining when to call setupModal or teardownModal.
+    // Elements are not yet available in the DOM due to `if (!visible) return null;`
+    // See componentDidUpdate.
+    this.isVisibleDirty = true;
+  }
+
+  componentDidLoad() {
+    if (this.visible) this.setupModal();
+  }
+
+  /**
+   *
+   * Stencil's componentDidUpdate doesn't provide us with previous props to compare
+   * and determine if we need to setup or destroy the modal. We can use a boolean
+   * variable inside a Watch decorator as a workaround to determine if an update needs
+   * to occur.
+   */
+  componentDidUpdate() {
+    if (!this.isVisibleDirty) return;
+
+    this.isVisibleDirty = false;
+    if (this.visible) {
+      this.setupModal();
+    } else {
+      this.teardownModal();
+    }
+  }
+
+  disconnectedCallback() {
+    this.teardownModal();
+  }
+
+  private handleClose(e) {
+    this.closeEvent.emit(e);
+  }
+
+  /**
+   * This method traps the focus inside our web component, prevents scrolling outside
+   * the modal, and adds aria-hidden="true" to all elements outside the web component.
+   * Fires analytics event unless disableAnalytics is true.
+   */
+  private setupModal() {
+    const initialFocus = (this.el.querySelector(this.initialFocusSelector) ||
+      this.el.shadowRoot.querySelector(
+        this.initialFocusSelector,
+      )) as HTMLElement;
```

Comment:
I think we would still need `isVisibleDirty` because we can't call `setupModal` or `teardownModal` until after `render`. We could do the `isVisibleDirty` check inside componentShouldUpdate, but Stencil docs state `...it's not good to rely on it to watch for prop changes, instead use the @Watch decorator for that.`

Reply:
The query selector is to allow developers to be able to focus elements in either light or shadow DOM. I can remove this functionality, but I think it adds a lot of value. As a developer using this component, I expect to focus the element I want when providing a value for `initialFocusSelector`.